### PR TITLE
[BEAM-57] Makes FileBasedSink use a temporary directory

### PR DESCRIPTION
**Pull request by jkff**
When writing to `/path/to/foo`, temporary files would be written to `/path/to/foo-temp-$uid` (or something like that), i.e. as siblings of the final output. That could lead to issues like http://stackoverflow.com/q/39822859/278042

Now, temporary files are written to a path like: `/path/to/temp-beam-foo-$date/$uid`. This way, the temporary files won't match the same glob as the final output files (even though they may still fail to be deleted due to eventual consistency issues).

### CODE REVIEW

**Comment by jkff** [comment]
R: @dhalperi

File: examples/java8/src/test/java/org/apache/beam/examples/MinimalWordCountJava8Test.java [hunk 1]
```diff
@@ -68,7 +68,7 @@ public void testMinimalWordCountJava8() throws Exception {
      .apply(MapElements
          .via((KV<String, Long> wordCount) -> wordCount.getKey() + ": " + wordCount.getValue())
          .withOutputType(TypeDescriptors.strings()))
-     .apply(TextIO.Write.to("gs://YOUR_OUTPUT_BUCKET/AND_OUTPUT_PREFIX"));
```
Comment (dhalperi) [inline examples/java8/src/test/java/org/apache/beam/examples/MinimalWordCountJava8Test.java]: revert unrelated change?

**Comment by dhalperi** [comment]
R: @lukecwik

**Comment by lukecwik** [comment]
Both subdirectories and siblings have issues but subdirectory seems to be the lesser of two evils as the fallback. There is also PipelineOptions.tempLocation but that may not work at all since the IOChannelFactory may not be equivalent or there will be other underlying storage differences like the GCS bucket may be part of a different GCP project.

In the referenced example of the person reading the pipeline in later, paths that are ** will still encompass the temp files. Sibling fails when the output location is a GCS bucket since you can't just create arbitrary siblings at that level or in a local file system the path is "/".

Does it make sense for the IOChannelFactory to own creating temp file locations?

**Comment by jkff** [comment]
@dhalperi:
* If we write to a temp directory inside the path, a user's glob like "**/*" will match the temp files - that's why I opted for a sibling, it seems cleaner in terms of isolation.
* I used Path.resolve on Path objects provided by the IOChannelFactory (Path is an interface), so it should be ok. And I, indeed, had to add sibling resolution methods to our IOChannelFactory's to make it work.

@lukecwik:
* Agreed that we can't use tempLocation for the reasons you listed.
* Agreed that this solution won't work when the user wants to write to a root path (raw GCS bucket or /something on a filesystem). Perhaps in this case we should _require_ them to provide a temp location, since there's no obvious safe choice the SDK can make for them?
* By "creating temp file locations", you mean "I want to write to _this_ path, give me a scratchpad for that" - i.e. move the generation of a temp location from FileBasedSink to IOChannelFactory?

**Comment by jkff** [comment]
Ping.

**Comment by lukecwik** [comment]
I think this is something we should bring back to the dev list since there isn't a clean solution either way and both have drawbacks.

**Comment by jkff** [comment]
Discussion happening on https://www.mail-archive.com/dev@beam.incubator.apache.org/msg01726.html

**Comment by jkff** [comment]
Rebased; per thread on beam-dev@ and per offline discussion with @dhalperi the approach the PR is taking is okay.

**Comment by jkff** [comment]
Failures look flaky (connection errors with apache repos).

File: examples/java8/src/test/java/org/apache/beam/examples/MinimalWordCountJava8Test.java (hunk 1 above)
Comment (dhalperi) [inline examples/java8/src/test/java/org/apache/beam/examples/MinimalWordCountJava8Test.java]: ping?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/gcsfs/GcsPath.java [hunk 2]
```diff
@@ -431,12 +431,14 @@ public GcsPath resolve(String other) {
...
+    GcsPath parent = getParent();
+    return (parent == null) ? other : parent.resolve(other);
   }
 
   @Override
   public Path resolveSibling(String other) {
-    throw new UnsupportedOperationException();
+    GcsPath parent = getParent();
+    return (parent == null) ? fromUri(other) : parent.resolve(other);
   }
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/gcsfs/GcsPath.java]: Done.

**Comment by dhalperi** [comment]
LGTM, merged.
